analyzers/tests/SonarAnalyzer.UnitTest/CFG/Roslyn/CfgAllPathValidatorTest.cs

```diff
@@ -0,0 +1,108 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using FluentAssertions;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using SonarAnalyzer.CFG.Roslyn;
+
+namespace SonarAnalyzer.UnitTest.CFG.Roslyn
+{
+    [TestClass]
+    public class CfgAllPathValidatorTest
+    {
+        [TestMethod]
+        public void ValidateCfgPaths()
+        {
+            const string code = @"
+public class Sample
+{
+    public int Method2(bool condition)
+    {
+        if (condition)
+            return 42;
+        else
+            return 1;
+    }
+}";
+            var cfg = TestHelper.CompileCfg(code);
+            /*
+             *           Entry 0
+             *             |
+             *             |
+             *           Block 1
+             *           /   \
+             *     Else /     \ WhenFalse
+             *         /       \
+             *     Block 2    Block 3
+             *        \        /
+             *         \      /
+             *          \    /
+             *           Exit 4
+             */
+            var validator = new CfgValidate(cfg, new List<int> { 3, 4 });
+            validator.CheckAllPaths().Should().BeTrue();
+
+            // only entry block is valid
+            validator = new CfgValidate(cfg, new List<int> { 1, 2, 3, 4 });
+            validator.CheckAllPaths().Should().BeTrue();
+
+            var nonEntryBlockValid = new NonEntryBlockValid(cfg);
+            nonEntryBlockValid.CheckAllPaths().Should().BeTrue();
+
+            var emptyValidator = new EmptyValidator(cfg);
+            emptyValidator.CheckAllPaths().Should().BeFalse();
+        }
+
+        private class EmptyValidator : CfgAllPathValidator
+        {
+            public EmptyValidator(ControlFlowGraph cfg) : base(cfg) { }
+
+            protected override bool IsValid(BasicBlock block) => false;
+
+            protected override bool IsInvalid(BasicBlock block) => false;
+        }
+
+        private class NonEntryBlockValid : CfgAllPathValidator
+        {
+            public NonEntryBlockValid(ControlFlowGraph cfg) : base(cfg) { }
+
+            protected override bool IsValid(BasicBlock block) => block.Ordinal > 0;
+
+            protected override bool IsInvalid(BasicBlock block) => false;
+        }
+
+        private class CfgValidate : CfgAllPathValidator
+        {
+            private readonly List<int> invalidBlocks;
+
+            public CfgValidate(ControlFlowGraph cfg, List<int> invalidBlocks) : base(cfg)
```

Comment:
We don't really need a list, we won't be adding items to it. This will simplify the invocation. the field could be array too.
```suggestion
            public CfgValidate(ControlFlowGraph cfg, params int[] invalidBlocks) : base(cfg)
```